[video] fix leading/trailing whitespaces in video details

xhaggi opened the pull request:
This is an attempt to fix the crap provided by metadata websites (extra spaces before and after actors/studios/etc.... or \t in actors).

@Montellese @MartijnKaijser @mkortstiege mind taking a look.

xhaggi committed:
[video] fix leading/trailing whitespaces in video details

mkortstiege commented:
That's pretty much what i had in mind. Thanks!

Paxxi commented:
I dont think this should be in our xml code. It shouldnt know or care about the data. Move this to a higher layer.

mkortstiege commented:
> I dont think this should be in our xml code. It shouldnt know or care about the data. Move this to a higher layer.

IMO that's not possible with the current design. Having that said, please let's focus on getting a solution for Isengard. This (and the schema bump and cleanup)  is currently blocking the release cycle.

xhaggi commented:
@Paxxi the trim stuff is optional and could be enabled via a parameter so why should we move it outside the `XMLUtils` and why do we name it `XMLUtils` if it doesn't provide such a convenience.

Montellese commented:
IMO this is ok for a start. It's certainly not the perfect place but it's the easiest for now. But we also have to keep in mind that it doesn't cover all the use cases. It is e.g. still possible to write crap into the database through JSON-RPC. The best and cleanest solution would be to introduce setters in `CVideoInfoTag` like `SetGenre()` which would then take care of the trimming etc.

Also remember that we still need special database migration handling for the crap that already is in the database.

Thanks for looking into it.

Paxxi commented:
@xhaggi it's optional sure but once you start sticking things in there that doesn't belong more quick fixes get stuffed in there and we end up with methods like we have now with 3-4 optional boolean parameters that nobody has a clue what they do unless they go read the implementation of that method

@mkortstiege quick and dirty hacks might be fine if there's a plan on a proper solution once Isengard is out the door but that's rarely the case and we end up in a worse shape for J***.

If @Montellese approves I won't object to it getting merged but I'd rather see a cleaner fix.

xhaggi commented:
@Montellese i've added setter's for all string members as you suggested and to show you how complicated things are, not because of the setter's but of the fact that `ParseNative()` can't use the members directly.

xhaggi commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -722,12 +773,14 @@ void CVideoInfoTag::ParseNative(const TiXmlElement* movie, bool prioritise)
         XMLUtils::GetInt(nodeDetail, "durationinseconds", p->m_iDuration);
-        XMLUtils::GetString(nodeDetail, "stereomode", p->m_strStereoMode);
+        XMLUtils::GetString(nodeDetail, "stereomode", value, true);
+        p->m_strStereoMode = StringUtils::Trim(value);
```
IMO in this cases it's really handy to say trim the value in `XMLUtils::GetString()`

xhaggi replied:
@Paxxi in some cases like this we don't have a setter which take care of the trimming, should we leaven the trim support in XMLUtils and use it here. It would be really handy in that cases.

xhaggi commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -674,13 +713,22 @@ void CVideoInfoTag::ParseNative(const TiXmlElement* movie, bool prioritise)
+
+  std::vector<std::string> studio(m_studio);
   XMLUtils::GetStringArray(movie, "studio", m_studio, prioritise, g_advancedSettings.m_videoItemSeparator);
+  SetStudio(studio);
```
same here, it's better to say trim the values in `XMLUtils::GetStringArray()`

xhaggi replied:
BTW use the handy trim support in XMLUtils + use the members directly will safe most of the lines.
The setter could be used outside, for now we can fix the JSON-RPC stuff by using the setter.

Montellese replied:
The current code won't work because you still use `m_studio` in the `GetStringArray()` call (same in other places).

xhaggi replied:
c/p ... will fix it

Paxxi commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -820,3 +874,168 @@ unsigned int CVideoInfoTag::GetDurationFromMinuteString(const std::string &runti
+  return StringUtils::Trim(v);
+}
+
+std::vector<std::string> CVideoInfoTag::Trim(const std::vector<std::string> &items)
```
make this one take std::vector<std::string>&& items) and then call it with std::move(param) and it'll be a zero copy operation

Paxxi commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -820,3 +874,168 @@ unsigned int CVideoInfoTag::GetDurationFromMinuteString(const std::string &runti
+
+std::vector<std::string> CVideoInfoTag::Trim(const std::vector<std::string> &items)
+{
+  std::vector<std::string> vecItems(items);
```
and remove this line

Paxxi commented:
:+1:

xhaggi commented:
@Montellese please let me know where to use the setter in the JSON-RPC part and I'll do the necessary changes tomorrow.

Tolriq commented:
For information this kind of problem also touches audio part, I have to do trims and cleanup on Yatse side, but I suppose this touch the tag parser and not this xml stuff.

xhaggi commented:
Addressed all comments. please take another look.

xhaggi committed:
[video] de-duplication/cleanup of related database tables

xhaggi commented:
@Montellese i've added a separate commit which trim the related columns in the video database. hope it's that what we need.

Paxxi commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -820,3 +869,168 @@ unsigned int CVideoInfoTag::GetDurationFromMinuteString(const std::string &runti
+  m_basePath = Trim(basePath);
+}
+
+void CVideoInfoTag::SetDirector(const std::vector<std::string> &director)
```
Make this non const and && otherwise std::move does nothing and we still copy the array, same for the other methods that take an array

Paxxi commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -820,3 +869,168 @@ unsigned int CVideoInfoTag::GetDurationFromMinuteString(const std::string &runti
+  m_strUniqueId = Trim(uniqueId);
+}
+
+std::string CVideoInfoTag::Trim(const std::string &value)
```
You can use the same idea as for arrays here to avoid the copies

Paxxi commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -820,3 +869,168 @@ unsigned int CVideoInfoTag::GetDurationFromMinuteString(const std::string &runti
+  return StringUtils::Trim(v);
+}
+
+std::vector<std::string> CVideoInfoTag::Trim(const std::vector<std::string>&& items)
```
Remove const and remove the assignment to values below, we want to get rid of copying and use move semantics as much as possible

Paxxi commented:
Passing parameters are starting to be seen as a form of antipattern and its something clang modernize warns about. You can read a bit more about why I'm going on about this here http://clang.llvm.org/extra/PassByValueTransform.html . Getting rid of it will be a process but it will end up with more natural code such as
```
Auto a = Trim(xmlutil::getvalue(abc));
```
With zero copying

xhaggi commented:
@Paxxi I totally agree added a fixup which drops the const and use it for both trim methods.

xhaggi committed:
[rfc][json-rpc] use setter of CVideoInfoTag

xhaggi commented:
@Montellese added another commit which refactor the json-rpc part to use the setter of `CVideoInfoTag`

Montellese commented:
The setters and changes to JSON-RPC look good but I'd drop the videodb commit as @MilhouseVH has already done a lot of work on that in #7048 which should be extended.

MartijnKaijser commented:
Perhaps pull that PR into this one so we can have an easy way to create test-builds and merge it all on one go?

xhaggi commented:
@Montellese how does #7048 fits if your database has already the new actor table structure? I've got a test-db `MyVideo92.db` which holds crap in the new `actor` table.

@MartijnKaijser sure if it make sense to pull the changes into this PR.

Montellese commented:
@xhaggi: Nobody said that #7048 was ready otherwise it would have already been merged. But it contains a lot of migration logic that is needed for people that move directly from Helix stable to Isengard stable (i.e. without installing any Isengard nightlies/betas/RCs) which IMO is the most important part of the update logic that absolutely needs to work.

xhaggi commented:
I never said something about ready, I want to know: how does it fits into this PR?. Why do you thing I should drop my db related changes?

xhaggi commented:
@Montellese i've added a new fixup commit which cleanup all related database tables and will be working for db's up to 92. It would be nice if you could take a look.

Montellese commented:
@xhaggi: We will need the full update logic in the update to version 91 as well because otherwise for databases with bad content the creation of the unique indeces will fail and they'll never make it to the update to 93. That's why #7048 modifies the update to version 91. But it's missing the handling for studio, genre, country and tag.

xhaggi commented:
@Montellese addressed all your comments. I'll move the changes from #7048 over to this PR and will do the necessary changes to studio etc.

xhaggi commented:
@Montellese is there a way we do a quick chat (e.g. IRC) about the necessary changes?

xhaggi commented:
I've added new commit which moves over the distinct stuff from #7048. Furthermore I've dropped the actor duplication migration from the 91's part because this is done by the code I've added. Now the cleanup (remove of duplicate items) is done from db version 90 till current.

MilhouseVH commented:
Quick test with this PR didn't go well: http://sprunge.us/WLNO

It worked with PR7048.

xhaggi commented:
@MilhouseVH could you please try another run. I've added a fix for that issue, tested only sqlite before.

MilhouseVH commented:
Testing again with 1bd55f2: http://sprunge.us/iFSB

It's now hanging the MySQL server with the following query:
```
SELECT (SELECT MIN(o2.actor_id) FROM actor o2 WHERE TRIM(o2.name) = TRIM(actor.name)), actor_id FROM actor WHERE TRIM(name) IN (SELECT TRIM(name) FROM actor GROUP BY TRIM(name) HAVING COUNT(1) > 1) AND actor_id <> (SELECT MIN(o2.actor_id) FROM actor o2 WHERE TRIM(o2.name) = TRIM(actor.name))
```
The MySQL server is at 100% CPU load and eventually I have to restart the service.

MySQL is running on an HP N36L with 8GB RAM so not the fastest box in the world. I'll run some tests, see if the query can be optimised...

xhaggi commented:
I'm sure the query dont use index scans. how many rows holds the table?

MilhouseVH commented:
`actor` has 40335 rows

xhaggi commented:
could you do a `EXPLAIN ...` and pastebin it.

MilhouseVH commented:
http://pastebin.com/raw.php?i=JQDQXCcj

xhaggi commented:
thanks,  as i said no index is used which is really slow for lots of rows with all the aggregations+sub selects

xhaggi commented:
best if i can get access to your mysql server to do the necessary optimizations. or if you are familiar with sql you could provide it?

MilhouseVH commented:
I can't really provide access to the server although I can send you the v90 database. Unfortunately I think the query is never going to work well as it stands, I'd imagine any reasonably sized SQLite `actor` table will choke on a Pi, are you testing this on a Pi or i7? :)

I'm not even sure what you're trying to do with this new code, to be honest... if it's a replacement for `CleanupActorLinkTablePre91`, why not go back to that? It's not fast, but at least it will complete eventually.

xhaggi commented:
it will cleanup all tables not only actor. if we can't fetch it by one query we can split it into two. I'll take a look at it tomorrow.

xhaggi commented:
@MilhouseVH I've optimized the query and it returns now ;) .. it would be great if you could run another test.

MilhouseVH commented:
Still a problem, I'm afraid, with the same SELECT() never finishing.

One solution, which is very quick and dirty but maybe acceptable for a migration, would to be to use a couple of temporary tables, that could be created, truncated, and re-used then finally dropped:
```
CREATE TABLE `temptable1` (
  `id` int(11) NOT NULL,
  `name` text,
  PRIMARY KEY (`id`),
  INDEX `ix_temptable1_name` (`name`(255))
)

CREATE TABLE `temptable2` (
  `name` text,
  UNIQUE KEY `ix_temptable2_name` (`name`(255))
)
```
Clear out before use:
```
TRUNCATE TABLE temptable1
TRUNCATE TABLE temptable2
```

Then populate the temp tables with the trimmed values (in this case, from `actor`):
```
INSERT INTO temptable1(id, name)
	SELECT actor_id, TRIM(name) FROM actor
```
and then the duplicate items:
```
INSERT INTO temptable2(name)
  SELECT name FROM temptable1 GROUP BY name HAVING COUNT(1) > 1
```
and now the query that never finishes becomes:
```
SELECT (SELECT MIN(o2.id) FROM temptable1 o2 WHERE name = temptable1.name) as primary_id, id as secondary_id FROM temptable1 WHERE name IN (SELECT name FROM temptable2)
```
(I've named the columns in the result which although not strictly necessary may make the query a little more understandable)

*or* an alternative version which I prefer, using `EXISTS` instead of `IN`:
```
SELECT (SELECT MIN(o2.id) FROM temptable1 o2 WHERE name = temptable1.name) as primary_id, id as secondary_id FROM temptable1 WHERE EXISTS (SELECT 1 FROM temptable2 WHERE name = temptable1.name)
```
(then when all data is migrated, drop `temptable1` and `temptable2`)

With the above changes, the never-ending query now completes sub-second (0.483s). The time taken to create the temporary tables is under 2 seconds (40335 rows inserted into `temptable1`, 97 rows inserted into `temptable2`).

This gives the following results (`name` column added, ordered by `primary_id`): http://pastebin.com/t6wwnFtu

Looking at my earlier [testing](https://github.com/xbmc/xbmc/pull/6020#issuecomment-68310625), it seems to have correctly identified these items, which is promising:
```
'22982', 'José Zúñiga'
'27647', 'Jose Zuniga'
'22882', 'Karel Dobrý'
'28142', 'Karel Dobry
```

I haven't however tested the above changes in Kodi, only in MySQL Workbench. Hopefully it will also work in SQLite.

What do you think?

xhaggi commented:
As a hint for all followers we found a solution for the MySQL issues.

MartijnKaijser commented:
Worked fore on a Frodo version upgrade. Sadly it wasn't legacy scan but a fresh one and then upgraded

MartijnKaijser commented:
If you missed the irc comments, android fails to build.

xhaggi committed:
[database][mysql] insert char set for CREATE TABLE .. SELECT syntax between create table and select clause

MilhouseVH commented:
Great news - with the commits here I've been able to migrate 3 MySQL databases that were previously a problem:

Database | Time | Stats | Debug log | Forum thread
---- | ---- | ----- | --------- | ------------
njm90 | 2m28s | 40,355 actors; 97 dupes | [link](http://sprunge.us/VHbd) |
bamzero90 | 0m52s | 22,087 actors; 8 dupes | [link](http://sprunge.us/QFQA) | [link](http://forum.kodi.tv/showthread.php?tid=223165&pid=2003003#pid2003003)
bossanova90 | 6m39s | 42,560 actors; 200 dupes | [link](http://sprunge.us/BMGb) | [link](http://forum.kodi.tv/showthread.php?tid=225859&pid=1995568#pid1995568)

The more duplicates, the slower the migration, unfortunately - being able to show a "please wait" splash (or similar) would be great!

We do still need to cherry pick https://github.com/MilhouseVH/xbmc/commit/807c1ddd1ec98b22c5264a530ddd40534e0348a5 from #7048 - this is needed because actors can be any length, however the unique index is only 255 chars, so anything anything longer than 255 chars will need to be truncated (to simplify subsequent de-duplication). The truncation of the column data needs to happen early in v93, prior to de-duplication. Also, new actors must only be inserted when the first 255 chars of the actor are unique otherwise `<same-255-chars>A` and `<same-255-chars>B` will fail the unique constraint.

Many thanks for all the hard work, @xhaggi!

I'll include this PR in my next RPi build if it's not already merged and no other issues are found.

xhaggi committed:
[video] limit length of name to 255 for actor, country, genre etc.

xhaggi commented:
@MilhouseVH thanks for your detailed comment :+1:

To your changes in https://github.com/MilhouseVH/xbmc/commit/807c1ddd1ec98b22c5264a530ddd40534e0348a5

The second part was already in for all related tables and was done in section 91 of `UpdateTables()`. Now I've move it into the 93 section see 7addef1 as we don't need to do it twice.

The first part of your changes should be addressed with d3b5847. We don't need to truncate the column because we it was already truncated to 255 in the migration part, but we need to truncate the value before insert. Without this it won't work and we already end-up with names greater 255. I've done this for all other related tables genre, studio etc. in `CVideoDatabase::AddToTable()`.

@MartijnKaijser the android build-error should be addressed with cf7d9ae. Could you start a new test-build please.

MartijnKaijser commented:
Builds all succeeded. They can be found on the server containing kodi-20150611-58a4db2

@MilhouseVH if you could test again i guess we're almost good to go. Only need some squashing.
perhaps @mkortstiege @Montellese or any others for a quick review

xhaggi commented:
I've pushed another commit which should optimize the migration process for the link tables. It drops the subquery in the update statement and also the mysql workaround. @MilhouseVH it would be nice if you could do some tests with the latest commit.

MilhouseVH commented:
With the latest updates and performance optimisations I've migrated the 3 test databases without any problem - timings are now much improved:

Database | Time | Stats | Debug log | Issue thread
---- | ---- | ----- | --------- | ------------
njm90 | 1m01s | 40,355 actors; 97 dupes | [link](http://sprunge.us/ifhd) | https://github.com/xbmc/xbmc/pull/6020#issuecomment-68048430
bamzero90 | 0m45s | 22,087 actors; 8 dupes | [link](http://sprunge.us/EQTW) | [link](http://forum.kodi.tv/showthread.php?tid=223165&pid=2003003#pid2003003)
bossanova90 | 1m34s | 42,560 actors; 200 dupes | [link](http://sprunge.us/IdFJ) | [link](http://forum.kodi.tv/showthread.php?tid=225859&pid=1995568#pid1995568)

Testing with MySQL 5.5.21/FreeNAS 8.3.x (FreeBSD) on an HP N36L (1.3GHz AMD Athlon II Neo dual-core) with 8GB RAM and a 64MB /tmp partition.

Hopefully this should give some confidence that migrations will succeed on modest systems - @xhaggi spent a lot of time optimising queries that would fill my /tmp partition.

I've also tested a Pi2 upgrading my current MyVideos92 MySQL db to MyVideos93, and that completed successfully in 30 seconds.

I'll include this PR in tonight's RPi test build, and unless something comes out of that and any other testing, IMHO this is ready to go.

phil65 commented:
just wanted to throw in a quick "Thank you!" for your efforts to solve this problem, @xhaggi and @MilhouseVH :)

MartijnKaijser commented:
indeed a big thank you :)
jenkins build this please. I think we should get this in asap

xhaggi commented:
Fixed the mistakes I've spotted after squashing everything together.

jenkins build this please

MartijnKaijser commented:
hold, some one broke master

xhaggi commented:
okay, found another one, now everything should be fine ;)

MartijnKaijser commented:
you sure ;)

xhaggi commented:
yep, tested with sqlite, 22,087 actors, found 22 dupes, needs ~6 sec. 
System: Intel i5, 4GB RAM, SSD.

MartijnKaijser commented:
jenkins build and merge :)

MilhouseVH commented:
I've just quickly re-tested the 3 databases after the latest changes, all good here.

hudokkow commented:
Nice!

Forage commented:
Could this be the cause of movies not getting added properly to the library any longer? See http://forum.kodi.tv/showthread.php?tid=224025&pid=2028063#pid2028063 for more info

justinkb commented:
Think this broke tvdb scraping (well, not the actual scraping but the processing of the xml data?) in some way... Getting blank titles on refreshing show information.

MartijnKaijser commented:
@justinkb pull master again as that's already fixed

justinkb commented:
I understand from the more recent commits it was supposed to be fixed, but I fear it's not since I posted this while on the latest revision... But I'll do another completely clean build to double check.

MilhouseVH commented:
@Forage has [confirmed](http://forum.kodi.tv/showthread.php?tid=224025&pid=2028546#pid2028546) his scraping issues are fixed by #7276.

justinkb commented:
@MilhouseVH excellent news. I must've pulled just before the fix went into master or something. Recompilation on my end is just about finished so I should be able to verify it as well in a few minutes
